### PR TITLE
maintainers/scripts/remove-old-aliases.py: script to remove aliases older than $date

### PR DESCRIPTION
**Pull request by Artturin**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
...but why `perl`?

**Comment by Artturin** [comment]
> ...but why `perl`?

is perl not a good language for this? i've just recently started using perl so i chose to write this in it.

i'll write this in python too as a learning experience

**Comment by Artturin** [comment]
wrote it in python,
here is the perl script for anyone curious
```perl
#!/usr/bin/env nix-shell
#!nix-shell -i perl -p perl perlPackages.ListMoreUtils perlPackages.RegexpCommontime
use strict;
use warnings;
use English;
use Cwd qw(abs_path);
use Carp qw(croak);
use Regexp::Common qw(time);
use List::MoreUtils qw(any);
use File::Copy qw(move);
use experimental 'switch';
use 5.034;
our $VERSION = 1; # PerlCritic wanted this

my $aliasfilename;
my $cutoffdate;
my $dryrun;

for (my $n = 0; $n < scalar @ARGV; $n++) {
    my $arg = $ARGV[$n];
    given ($arg) {
        when ( $_ eq "--year") {
            $n++;
            $cutoffdate = $ARGV[$n];
            croak "--year’ requires an argument" unless defined $cutoffdate;
        };
        when ( $_ eq "--file") {
            $n++;
            $aliasfilename = abs_path($ARGV[$n]);
            croak "--file’ requires an argument" unless defined $aliasfilename;
        };
        when ( $_ eq "--dry-run") {
            $dryrun = 1;
        };
        default { croak "usage ex: --year 2020 --file aliases.nix | optional --dry-run"; }
    };
};

croak "no filename" if not defined $aliasfilename;
croak "no year" if not defined $cutoffdate;
my @dateline;
my @onlyrow;
my @separateline;
my @sorted0;
my @sorted1;

sub getdatelines {
    open my $fh, '<:', $aliasfilename
        or croak "Could not open file $aliasfilename $ERRNO";
    while (my $row = <$fh>) {
        chomp $row;
        if ( $row =~ /=.+$RE{time}{ymd}/sm ) {
            my @dates = $row =~ /$RE{time}{ymd}{-keep}/smx;
            my $date = "$dates[0]";
            if ($date lt $cutoffdate) {
                push @dateline, "$date | $INPUT_LINE_NUMBER | $row";
                push @onlyrow, $row;
            };
        };
        # dates that are not on the same line as the assignment
        if ( $row =~ /^(?:(?!=).)*$RE{time}{ymd}/smx ) {
            my @dates = $row =~ /$RE{time}{ymd}{-keep}/smx;
            my $date = "$dates[0]";
            if ($date lt $cutoffdate) {
                push @separateline, "$date | $INPUT_LINE_NUMBER | $row";
            };
        };
    };

    close $fh or croak "Could not close $aliasfilename $ERRNO";

    push @sorted0, sort @dateline;
    push @sorted1, sort @separateline;
    foreach (@sorted0) {
        say;
    };
    if (@sorted1) {
        say "========== on separate line, remove or fix manually ==========";
        foreach (@sorted1) {
            say;
        };
    };
    return;
};
getdatelines();

sub writefile {
    open my $fh, '<:', $aliasfilename
        or croak "Could not open file $aliasfilename $ERRNO";
    open my $writefh, '>:', "$aliasfilename.raliases"
        or croak "Could not open file $aliasfilename.raliases $ERRNO";
    while (my $row = <$fh>) {
        chomp $row;
        my $quoted = quotemeta "$row";
        # filter the dates older than $cutoffdate out
        if (!any { /^$quoted/sm } @onlyrow) {
            say {$writefh} $row;
        } elsif ($row =~ /^\s*$/sm) {
            # ^ if row is just a newline. for some reason the say above does not print those.
            say {$writefh} "$row";
        };
    };

    close $fh or croak "Could not close $aliasfilename $ERRNO";
    close $writefh or croak "Could not close $aliasfilename.raliases $ERRNO";

    move("$aliasfilename.raliases", $aliasfilename) or croak "Move operation failed $ERRNO";
    return;
};

writefile() unless $dryrun;
```

**Comment by aanderse** [comment]
Cool. `python` seems like a much better choice to me. I think we allow `python` for scripts like this 👍

File: maintainers/scripts/remove-old-aliases.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i python3 -p "python3.withPackages(ps: with ps; [ dateutil ])"
+# example usage:
+# $ NIX_PATH=nixpkgs=$PWD ./maintainers/scripts/remove-old-aliases.py --year 2018 --file ./pkgs/top-level/aliases.nix --dry-run
+# Also check this file with mypy after every change!
```
Comment (Mic92) [inline maintainers/scripts/remove-old-aliases.py:5]: I removed all globals and added some docs.

Comment (Artturin) [reply maintainers/scripts/remove-old-aliases.py:5]: thanks

**Comment by Artturin** [comment]
should we convert them to throws first?

**Comment by stigtsp** [comment]
> > ...but why `perl`?
 
> is perl not a good language for this? i've just recently started using perl so i chose to write this in it.

Sorry @Artturin I unfortunately missed this PR eariler, so a bit late to the party. I've added some comments on the initial script below.

@aanderse Imho: Perl and Python are both equally good languages for this kind of task. I don't think we should dissuade contributions in a mature systems language unless there are specific reasons for it.

```perl
...
### croak() and friends are mostly useful when writing modules, as errors will be
### in the context of the caller. https://perldoc.perl.org/Carp
###
use Carp qw(croak);
...
### Low-severity perlcritic can be a bit too picky, imho. A $VERSION is
### certainly needed in a vendored module, but imho not really useful in scripts.
###
our $VERSION = 1; # PerlCritic wanted this
...
### Consider using using Getopt::Long (or something more modern like Applify) to
### automatically handle input arguments. The latter also automatically
### usage doc based on expected input arguments.
###
for (my $n = 0; $n < scalar @ARGV; $n++) {
    my $arg = $ARGV[$n];
...
sub getdatelines {
    open my $fh, '<:', $aliasfilename
      or croak "Could not open file $aliasfilename $ERRNO";
    ### chomp() can be used in the while condition directly, but something like
    ###
    ###    use File::Slurp;
    ###    my (@rows) = read_file($aliasfilename);
    ###
    ### .. is perhaps easier to read, and handles exceptions automatically.
    ### Mojo::File is also a nice module to use when dealing with most file
    ### operations.
    ###
    while (my $row = <$fh>) {
        chomp $row;
        if ( $row =~ /=.+$RE{time}{ymd}/sm ) {
          ### You can assign $date directly in the my by setting it up in list
          ### context like this:
          ###
          ###   my ($date) = $row =~ m/=.($RE{time}{ymd}{-keep})/smx;
          ###
            my @dates = $row =~ /$RE{time}{ymd}{-keep}/smx;
```

**Review by stigtsp — APPROVED** [review]
LGTM
- Tested OK, cleans up aliases as expected

**Comment by aanderse** [comment]
> unless there are specific reasons for it.

To the best of my knowledge there are approximately 4 people active in our community that are comfortable writing `perl`.

If I had to make a wild guess I would exaggerate to make a joke and illustrate a point at the same time: I would guess there are maybe 4 people in our community who are **not** willing to write `python`.

It isn't that there is something intrinsically wrong with `perl`, but purely from a demographics point of view our community is not overly willing or able to maintain `perl` code. Look at many PRs showing people's hesitancy to modify any of our`perl` scripts or hack on `hydra`

I understand you're very capable with `perl` and I **greatly** appreciate that about you. I have had to be responsible for many programs written in `perl`. Thanks to you and the `perl` contributors in this community I was able to confidently port these applications to NixOS which greatly improved my life. I don't mean to trash talk what you do... I just try to make a relevant point to maximize supportability within our community.

As always thanks for being so great @stigtsp ❤️

**Comment by Artturin** [comment]
now the script first converts old aliases to throws and adds a comment "Converted to throw $date"
```diff
- bridge_utils = bridge-utils; # Added 2015-02-20
+ bridge_utils = throw "'bridge_utils' has been renamed to/replaced by 'bridge-utils'"; # Converted to throw 2022-02-20
```

once a throw has a date older than the cutoff date , the throw line will be removed.

i still have to add some checks and clean the script a bit and then its ready

**Comment by Artturin** [comment]
alright, ready.

**Comment by piegamesde** [comment]
In general I am in favor of doing this, but I'm really uncomfortable with the idea of stream editing Nix code like this, especially if we later want to run this automatically without human attendance.

Therefore, I'd like to evaluate the possibility of migrating our aliases to a machine-readable format (JSON, YAML, TOML, whatever) so that our script can safely run there.

**Comment by Artturin** [comment]
ready, and added a syntax check with `nix-instantiate --eval`

**Comment by Artturin** [comment]
> In general I am in favor of doing this, but I'm really uncomfortable with the idea of stream editing Nix code like this, especially if we later want to run this automatically without human attendance.
> 
> Therefore, I'd like to evaluate the possibility of migrating our aliases to a machine-readable format (JSON, YAML, TOML, whatever) so that our script can safely run 

i'd prefer a json file imported with `importJSON`

```json
{
    "letsencrypt": {
        "points-to": "certbot",
        "added": "2016-05-14",
        "expiry": "2018-01-01",
        "throw": false,
    }
}
```

however if we had to do throw logic withing evaluation then the nixpkgs eval would become slower right?

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/161146

**Comment by piegamesde** [comment]
It is sad that we restrain ourselves from implementing solutions we deem better because of fear of performance regressions (this is not the first time I see this). Obviously the "correct" way would be to implement both and then measure the performance impact to determine if it is worth it. Of course, this is not a practical approach because it requires more than double the amount of work. :(

**Review by stigtsp — APPROVED** [review]
LGTM
- Works OK, tested with `maintainers/scripts/remove-old-aliases.py --year 2019 --month 6 --file ./pkgs/top-level/aliases.nix`

**Pull request merged**